File: P_4/Examen_P4/Napakalaki_Ruby_P4/lib/cultist_player.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

class CultistPlayer < Player
   
    @@totalCultistPlayers = 0
    
    attr_reader :totalCultistPlayers    
    
    def initialize(p, c)
        Player.crear_copia(p)
        @myCultistCard = c
        @@totalCultistPlayers = @@totalCultistPlayers + 1
    end
    
    protected
    def get_combat_level
        
        combat_level = (super.get_combat_level * 1.2) +
                    (@myCultistCard.get_gained_levels * @@totalCultistPlayers)
        
        combat_level_truncado = combat_level.truncate
        
        return combat_level_truncado
    end
    
    def get_oponent_level(m)
        return m.get_combat_level_against_cultist_player
    end
    
    def give_me_a_treasure  
        
        tVisibles = @visible_treasures
    
        rnd = rand(tVisibles.length)
    
        return (tVisibles.delete_at(rnd))
    end
    
    def should_convert
        return false
    end
    
    def can_you_give_me_a_treasure
        
        t = @enemy.visible_treasures
        have_treasures = !(t.empty?)
        
        return have_treasures
    end
end

File: P_4/Examen_P4/Napakalaki_Java_P4/src/napakalaki/CultistPlayer.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalaki;

import java.util.ArrayList;
import java.util.Random;

/**
 *
 * @author ixjosemi
 */

// Subclase de la clase player obtenida a partir de herencia
public class CultistPlayer extends Player{
    
    Cultist myCultistCard;
    static int totalCultistPlayers = 0;
    
    // Constructor
    public CultistPlayer(Player p, Cultist c) {

        super(p); // Llamada al constructor de copia de Player.
        totalCultistPlayers++; 
        myCultistCard = c;
    }
    
    // Devuelve el numero total de sectarios (jugadores)
    public static int getTotalCultistPlayers(){
        return totalCultistPlayers;
    }
    
    // Calcula el nivel que tiene el jugador sectario
    int getCombatLevels(){
        double combatLevel = (super.getCombatLevel()*1.2)+
            (myCultistCard.getGainedLevels()*totalCultistPlayers);
        
        int combatLevel_truncado = (int)combatLevel;
        
        return combatLevel_truncado;
    }
    
    // Devuelve el monstruo del oponente del jugador sectario
    protected int getOponentLevel(Monster m){
        return (m.getCombatLevelAgainstCultistPlayer());
    }
    
    // Devuelve un tesoro al azar de los tesoros visibles
    protected Treasure giveMeATreasure() {
        ArrayList<Treasure> tVisibles = this.getVisibleTreasures();
        
        Random r = new Random ();       
        int i = r.nextInt(tVisibles.size());
        
        return (tVisibles.remove(i)); 
    }

    // Devuelve siempre "false" ya que un sectario no puede volver a convertirse
    // en sectario
    protected boolean shouldConvert(){
        return (false);
    }
 
    // Devuelve si el enemigo tiene tesoros visibles que puedan robarse
    protected boolean canYouGiveMeATreasure(){
        
        Player enemy = this.getEnemy();
        
        ArrayList<Treasure> t = enemy.getVisibleTreasures();

        boolean tiene = !(t.isEmpty());
        
        return (tiene);
    }
 
    public String IsCultist(){
        return "Sí";
    }
}
